velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -202,310 +195,218 @@ inline static TResult addLarge(
+      // zeros.
+      int128_t aRescaled = a * velox::DecimalUtil::kPowersOfTen[aRescale];
+      int128_t bRescaled = b * velox::DecimalUtil::kPowersOfTen[bRescale];
+      r = reduceScale(
+          TResult(aRescaled + bRescaled), std::max(aScale, bScale) - rScale);
     } else {
-      // Fast path if one or more arguments are encoded.
-      exec::DecodedArgs decodedArgs(rows, args, context);
-      auto a = decodedArgs.at(0);
-      auto b = decodedArgs.at(1);
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            a->valueAt<A>(row),
-            b->valueAt<B>(row),
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
+      // The risk of overflow should be considered. Add whole and fraction
+      // parts separately, and then combine.
+      r = addLarge<TResult, A, B>(a, b, aScale, bScale, rScale, overflow);
     }
   }
+}
 
- private:
-  R* prepareResults(
-      const SelectivityVector& rows,
-      const TypePtr& resultType,
-      exec::EvalCtx& context,
-      VectorPtr& result) const {
-    context.ensureWritable(rows, resultType, result);
-    result->clearNulls(rows);
-    return result->asUnchecked<FlatVector<R>>()->mutableRawValues();
-  }
+// Compute the add and subtract operation result precision and scale following
+// Hive's formulas.
+std::pair<uint8_t, uint8_t> computeAddSubtractResultPrecisionScale(
+    uint8_t aPrecision,
+    uint8_t aScale,
+    uint8_t bPrecision,
+    uint8_t bScale) {
+  auto precision = std::max(aPrecision - aScale, bPrecision - bScale) +
+      std::max(aScale, bScale) + 1;
+  auto scale = std::max(aScale, bScale);
+  return sparksql::DecimalUtil::adjustPrecisionScale(precision, scale);
+}
 
-  const uint8_t aRescale_;
-  const uint8_t bRescale_;
-  const uint8_t aPrecision_;
-  const uint8_t aScale_;
-  const uint8_t bPrecision_;
-  const uint8_t bScale_;
-  const uint8_t rPrecision_;
-  const uint8_t rScale_;
-};
+template <typename TExec>
+struct DecimalAddFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  template <typename A, typename B>
+  void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& /*config*/,
+      A* /*a*/,
+      B* /*b*/) {
+    auto aType = inputTypes[0];
+    auto bType = inputTypes[1];
+    uint8_t aPrecision = getDecimalPrecisionScale(*aType).first;
+    uint8_t bPrecision = getDecimalPrecisionScale(*bType).first;
+    aScale_ = getDecimalPrecisionScale(*aType).second;
+    bScale_ = getDecimalPrecisionScale(*bType).second;
```

Comment:
The process to get input precisions and scales are very similar for the four operations. Any chance to reuse this logic?

Reply:
It sets the struct member, so we cannot reuse it.

Reply:
Perhaps add a function to return the precisions and scales for the callers to use.

Reply:
getDecimalPrecisionScale returns a temp std::pair<precision, scale>, it cannot directly set aScale_.
It will be 
auto [aPrecision, bPrecision, aScale, bScale] = getPrecisionAndScale.
aScale_ = aScale.
Do you think it is acceptable?
Or setPrecisionAndScale([aPrecision, bPrecision, aScale, bScale)?